### ISSUE TITLE
strange reconnect error

### ISSUE THREAD

**Issue by reharik** [nicdex/node-eventstore-client#60]
Hi,
I have an end to end test.  First it kills the ES instance and restarts it so it's a clean instance. 
So sometimes it reconnects and my handlers all say 'Caught up with previously stored events. Listening for new events.' And all is good, sometimes I get to that point and it doesn't pick up an new events form ES. 
The sequence is as follows

good case:
1) TcpPackageConnection: connection [{"host":"172.18.0.2","port":1113}, L{"host":"172.18.0.4","port":39716}, d5ae7be2-d5c2-47d1-b590-cbac22cfdb5f] was closed with error: Error: transmission error.. |
2) Subscription dropped.
3) Error: Connection was closed
4) TcpPackageConnection: connection to ...blah... ECOONREFUSED
5) TcpPackageConnection: connected to ...blah (success)
6) Caught up with previously stored events. Listening for new events.

bad case:
1) TcpPackageConnection: connection [{"host":"172.18.0.2","port":1113}, L{"host":"172.18.0.4","port":39716}, d5ae7be2-d5c2-47d1-b590-cbac22cfdb5f] was closed with error: Error: transmission error.. |
***No Subscription dropped. or Error: Connection was closed ***
2) TcpPackageConnection: connection to ...blah... ECOONREFUSED
3) TcpPackageConnection: connected to ...blah (success)
4) EventStoreConnection 'ES-87aa65ee-065c-4a97-a551-9cad38c8c046': operation never got response from server.
UTC now: Sun Oct 14 2018 17:35:25 GMT+0000 (Coordinated Universal Time), operation: Operation ReadAllEventsForwardOperation (dc12a6a4-6173-423a-a7c6-5bc51b738be9): Position: {"commitPosition":{"low":### **171680**,"high":0,"unsigned":false},"preparePosition":{"low":0,"high":0,"unsigned":false}}, MaxCount: 500, ResolveLinkTos: false, RequireMaster: true, retry count: 0, created: 17:35:18.153, last updated: 17:35:18.153. | 5:35:25 pm
5) Caught up with previously stored events. Listening for new events.
and then just hangs.

So in the bad case the subscriptions don't drop, and I get an error trying to connect with a commit position that is way higher than the latest event. I don't know where it's getting that number from but it's always the same and I believe it's the culprit.
Does this make any sense to you? I can't figure out what the heck is happening.
Thanks,
Raif

**Comment by nicdex** [comment]
Sorry I'm trying to understand what you're doing and I'm not sure I'm getting this right.

You have an end-to-end tests that kills an eventstore instance while there is a subscription connected and subscribed to that instance, using the nodejs client, and sometimes the subscriptions doesn't continue receiving new events after the eventstore is backup and sometime it does. Is that right?

Also one clarification question, that test eventstore service you're killing is it running with --memdb or with data stored on disk?

**Comment by reharik** [comment]
So if client tries to connect 10 times and fails it exits and calls my onClosed event which kills the process and pm2 restarts it and then were all good.  However, if it manages to connect before the 10th time then all my listeners report that they are caughtup and listening for new events.  However they don't pick up any of the new events.
The ES is on disk.
Unfortunately, it's a little difficult for me to debug, as the client, of course, is an imported module, and my processes are all in containers and the whole killing restarting thing is based on that.  So I can't change the node_modules code that is running inside the container. 
If I could I'd put a console log on "EventStoreConnectionLogicHandler.prototype._handleTcpPackage" and start tracing from there.
Thanks,
R

**Comment by reharik** [comment]
As a work around I tried to reduce the maxRetries:and maxReconnections hoping to basically force it to fail each time, but seems that reducing those numbers doesn't affect how many times it tries. To be specific, it fails and closes then it fails to connect 10 times. or it successfully connects.  I was hoping that making maxRetries like 5 then I'd only see 5 failed connections.
Sorry this is a mess, I'm just trying to do a brain dump. 
These three different states (closed, failed x 10, connected) are all logging from the last three functions in "TcpConnection.createConnectingConnection"  so lines 49-64 in tcpPackageConnection.
My point being that those functions are what are generating the "closed", "failed", "connected to" logs.

**Comment by reharik** [comment]
right so never mind about the MacReconnections, ... user error.  :)
Hopefully that will let me have a work around, but it does seem like I should be able to reconnect the eventstore if it comes back on line with out rebooting the process.

**Comment by reharik** [comment]
Oh great!  That's great to hear.  I mean not that there's a bug but that I'm only partially crazy :).  My subscriptions in question are all catch up variety.  So behavior is duly noted.
Thanks for your help!  I look forward to the fix.
Raif

**Comment by nicdex** [comment]
Bug fix is released. 

See https://github.com/nicdex/node-eventstore-client/issues/22#issuecomment-431708272 which also contains explanation of the subscriptionDropped / reconnect behavior.

I hope this solves the issues you had.

**Comment by reharik** [comment]
Hi, thanks for that fix.  It definitely seems to be trying to reconnect for all dropped subscriptions now.  However, what I'm getting is, if I have say 6 subscriptions, I get six errors like this

`EventStoreConnection 'ES-87aa65ee-065c-4a97-a551-9cad38c8c046': operation never got response from server.
UTC now: Sun Oct 14 2018 17:35:25 GMT+0000 (Coordinated Universal Time), operation: Operation ReadAllEventsForwardOperation (dc12a6a4-6173-423a-a7c6-5bc51b738be9): Position: {"commitPosition":{"low":### 171680,"high":0,"unsigned":false},"preparePosition":{"low":0,"high":0,"unsigned":false}}, MaxCount: 500, ResolveLinkTos: false, RequireMaster: true, retry count: 0, created: 17:35:18.153, last updated: 17:35:18.153. | 5:35:25 pm`

The key part being the event CommitPosition which is 171680.  This is way higher than the current event position.  Further more it's always that number: 171680 which makes me think it's so how hard coded somewhere.  If I get time away from the kiddo I'll try and dig in, but we have sun for the first time in weeks today, so I'm not sure.
Thanks,
r

**Comment by reharik** [comment]
omg, so I'm not seeing that damn error anymore.  But what I am seeing is that the liveProcessingStarted function fires for all the handlers and then it doesn't read any new events.  So then I start firing a bunch of events manually from the GES dashboard and eventually it picks up.  meaning, it seems to be waiting for a certain number of events to pass and then it starts listening.  Very strange.
I wish there was a GD way to see the position of an event from the dashboard.  That would really help me diagnose this.  I'm not sure where I would look in the node-eventstore-client, but there must be somewhere where an error is thrown because there is no connection, then that starts the retry process, then that is successful and somewhere it decides what the last processed event position is.  Somehow, I think, that is getting messed up.

One additional note is that I'm using the subscribeToAllFrom function.
Anyway, thanks for your help, I hope this makes some sense.
R

**Comment by reharik** [comment]
Right, of course it's not hard coded, that's the sound of a deranged and desperate mind.
Anyway, I'm going to have find a way to do some logging on this, it would be super helpful of you could let me know what classes I should look at.
I'm thinking that I need to look at wherever tries to reconnect the subscription after it's been dropped.  So somewhere it must say "ok try and start a subscription from this position". If that is actually just same code that starts it the first time, then maybe shed some light on how it's deciding where to start from.  I know the first time I'm passing in the position, so if it's the same code, then some code must be doing the same thing only with a new position.
This would be super helpful.
Thanks,
R

**Comment by AlexandruSimandi** [comment]
I am having this problem in 0.2.8, doesn't happen in 0.2.7

**Comment by nicdex** [comment]
Sorry but I think you have a false positive. The changeset between 0.2.7 and 0.2.8 is this only commit https://github.com/nicdex/node-eventstore-client/commit/e9d86a3c4d07ed91944bfc38b212716d6d72460f which only add an export for PersistentSubscriptionNakEventAction.

I'm having trouble reproducing this issue, can you guys send me a code example that reproduce this.

Also I need the a description of the environment you're using. I need the versions of node-eventstore-client (use yarn list or npm ls to get the actual version on disk), EventStore, node, the OS you are using, anything else you think I might need to reproduce this.

Thanks

**Comment by AlexandruSimandi** [comment]
@nicdex You are right, it was due to eventstore cluster, scaled to 1 and issue disappeared

**Comment by AlexandruSimandi** [comment]
@nicdex Is there any significant difference between 0.2.4 and 0.2.8 that might cause this? i would have checked that myself but the releases aren't tagged properly

**Comment by Raithybabes** [comment]
Hi.

Thanks for creating and maintaining this library @nicdex . I think this is the only viable Node port of the ES Client. I hope you are able to continue this great work until such time as there is an official port.

That said, I have hit an issue. I believe I am seeing the same as the original issue posted here, so I shall continue in this one rather than raise a new one.

I initially believed that my expectations were wrong, but your answers to similar issues posted here now lead me to believe that this is not working as intended. Your assistance would be much appreciated...

## Expected behaviour

When connection of a catch-up subscription drops for any reason, the client attempts to reconnect and, when reconnected, resume a working state (receiving messages).

I believe that existing connection event handlers should resume triggering - with no need to re-attach them.

## Known exemptions

I understand from other issues that this will only work for:

- catch-up subscriptions
- only once live processing has started

## Environment

- OS = Windows 10 1809
- Node = v11.9.0
- EventStore = 4.1.1.0 running from official Docker Image, with default settings

```
docker run --name myes -dt -p 2113:2113 -p 1113:1113 -e EVENTSTORE_DISABLE_HTTP_CACHING=true eventstore/eventstore
```

## My test approach

I am starting with the "subscribe-all-from-events.js" sample file. I have made some minor changes. My amended version is attached and all of my changes are marked with `// RWM:`.

[samp-subscribe-all-from-events.txt](https://github.com/nicdex/node-eventstore-client/files/2845181/samp-subscribe-all-from-events.txt)

Changes:
- Slowed down the `reconnectionDelay` to 1000 (1 second per attempt).
- Handled the `on('reconnecting ')` event to improve visibility of what is happening.
- `subscribeToStreamFrom` to subscribe to single stream "test" rather than all.

I am creating events directly into my "test" stream using the [ES web ui](http://localhost:2113/web/index.html#/streams/test/addEvent)

I am simulating connection drop by stopping the EventStore Container `docker stop myes`.

I am restoring the connection by starting the Container `docker start myes`.

## Actual behaviour

- The node app successfully connects to ES and catches up with past events, triggering `eventAppeared` handler.
- Triggers `liveProcessingStarted` handler when past events are caught up.
- Triggers `heartbeatInfo` handler as expected.
- I stop ES.
- Triggers `subscriptionDropped` handler, reason: "connectionClosed", error: "Error: Connection was closed".
- Triggers `reconnecting` handler 10 times, at 1 second intervals.

If I do NOT intervene:

- After 10 reconnecting attempts it gives up.
- Triggers `closed` handler, reason: "Reconnection limit reached.".
- The Node app exits.
- This is expected.

If I DO intervene:

- Before the reconnection limit is reached...
- I start ES.
- The `reconnecting` handler ceases tiggering.
- No other handled event is triggered.
- The Node app does NOT exit.
- The appearance is that it has recovered the connection, HOWEVER...
- No further events are received on: `heartbeatInfo`, `eventAppeared`, etc.
- I stop ES.
- No further events are received on: `subscriptionDropped`, `error`, `closed`, `reconnecting`, etc.
- The Node app requires closing with CTRL+C

I am nore sure what state the connection is in at this point and have not debugged deeper.

## Conclusion

This is where I thought it better to ask what the behaviour is actually supposed to be in this scenario.

Perhaps I *am* supposed to reconnect my event handlers?

**Please can you explain what I am doing wrong?**

Many thanks,
Raith

---

**Comment by nicdex** [comment]
@Raithybabes I haven't had time to investigate deeper but what I found out is that your issue is related to using eventstore in docker. 

I've tried the "intervene steps" with eventstore running locally on my machine and it works. The events handlers are triggered after the reconnect. 

Then I tried the same thing running eventstore in docker with the command you provided and I experience the same issue you have: the events handlers are not triggered after reconnect. 

I will investigate more because I am intrigued about this, but at the same time usage of eventstore in docker is out of the scope for this project so I might not have a fix for this.

**Comment by Raithybabes** [comment]
Thanks Nic. That is quite intriguing. I have become rather blasé regarding Docker magic and didn't consider the extra layers that the networking passes through. I can only assume that Docker applies a distinction between disconnected/closed/something-or-other that doesn't quite reflect the real world. Strange though that the Node end knows that it is no longer reconnecting (those events cease), and also doesn't consider the connection closed (that event doesn't trigger), yet gets left in a limbo state.

This is a blow, as we are rather dependent on Docker to support our development environment. We haven't got to Production yet but were expecting it to make an appearance there too. I hope this isn't a show-stopper!

If you can shed any more light on this then that'd be great but I understand that it's out of scope of your intended use case. Given that you've already identified it as a Docker issue, I now expect there'll be lots of similar stories when I google it from that angle...

**Comment by Raithybabes** [comment]
It's a very different scenario, but this strikes a chord.

https://github.com/moby/moby/issues/31208

> The issue is related to the way docker overlay routes packets, using first iptables to mark them and use ipvs to forward them to the right hosts but the default expiration for connections on ipvs is set to 900 seconds (`ipvsadm -l --timeout`) after which it will stop forwarding packets even though the connection still exists; If this happens then any new packet on this connection will now try to go to the virtual IP for that service that has no valid resolution, resulting in a broken state where it is stuck in limbo while the kernel forever tries to resolve that virtual IP.

This refers to a connection timing out, but *maybe* a similar thing happens with a dropped connection? It can be reconnected but the packet routing is broken by then??

It sounds like a Linux-y Docker issue that will only manifest in some implementations/OSes. I don't know if this is still an issue in current Docker. This particular issue was closed on the strength of the workarounds. This is beyond my Linux + Docker experience level.

@nicdex please can you elaborate on what takes place within the reconnection stage? Is the Node-end tcp connection closed at that point, or in an intermediate state? How does it negotiate the reconnect with the EventStore-end?

I have coded around this issue in my client by considering the connection closed and re-subscribing. The new subscription works fine, so there is no fundamental block to comms between Node and ES. It only affects the previous Connection object.

**Comment by nicdex** [comment]
It seems indeed to be a weird issue with tcp connections. 
If you look at the logs of the eventstore server and the client (you can activate logging by passing {log: console} to the createConnection function). 
You can see that the client says "TCP connection to ... established", but on the server side there is no "External TCP connection accepted" in the logs.

Weirdly enough in some cases when I activate the verbose logging (add {verboseLogging:true} to the settings passed to createConnection) it does work... so it seems to be an issue of timing after the tcp connection is established (because verbose logging adds some extra cpu cycles for doing the logging). So when connecting to a port opened for docker, it seems that the "connect" state is lying, socket doesn't seem to be ready to receive data immediately, but it doesn't throw any errors.

This is an intriguing one... I will continue to investigate.

**Comment by nicdex** [comment]
Fixed in 0.2.9 please try it.

**Comment by Raithybabes** [comment]
Hi Nic. I was away and missed v0.2.9 but just upgraded to 0.2.10 and the problem that I was seeing has indeed been fixed :)

Reconnection to ES running within Docker does now work using code which previously failed.

Whether or not this also fixes the original issue from @reharik will need confirming. Our issues sounded similar to me, right up until you deduced mine was related to Docker. From your fix code, I *guess* it will have a positive impact on the scenario that reharik detailed...?

**Comment by nicdex** [comment]
@reharik Can you check if 0.2.10 fixes your problem too, if so we can close this issue

**Comment by reharik** [comment]
I will do my best, it's kind of a process, I have to upgrade the project,
then upgrade rebuild all the images that use that project and so on and so
on.  I"m starting that process now.  Then I'll have to figure out what the
problem was cuz it's been a while.  But I'm on it.
Thanks,
Raif

**Comment by reharik** [comment]
Hmmm, well, I don't know about the previous error but I'm getting this

Error during processing ReadAllEventsForward request.
Log record at actual pos 36784 has too large length: 1416454501 bytes, while limit is 16777216 bytes. In chunk #0-0 (chunk-000000.000000).

I can't be certain that this happened as a result of the upgrade, but it seems likely.  I have no idea how to look at event at pos 36784 to see what it even is talking about.  I guess i'll try to roll back to see if this is the upgrade or something else.
R

**Comment by reharik** [comment]
Right.  Pollogies. User error. As I mentioned my setup is pretty complex. I missed a step. So.... seems to work, that's a base line. Now I have to try and remember how I worked around original problem so I can reproduce and see if .10 fixes it.  Might be a bit.  But it's moving along slowly.

**Comment by jeffijoe** [comment]
I thought I'd try to help out by running a basic test myself.

* Docker Compose with latest version of EventStore image
* Latest version of `node-eventstore-client`

I configured a script that subscribes to all, then pumped messages through it. I left that process, putting my Mac to sleep even. I came back the next morning, started the publishing process and the subscriber was still alive and received the messages instantly.

So at least under this fairly simple setup, everything seems good. 😄
